File: src/CSharp/CodeCracker/Refactoring/ParameterRefactoryCodeFixProvider.cs
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CodeActions;
using Microsoft.CodeAnalysis.CodeFixes;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.Formatting;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Composition;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

namespace CodeCracker.CSharp.Refactoring
{
    [ExportCodeFixProvider(LanguageNames.CSharp, Name = nameof(ParameterRefactoryCodeFixProvider)), Shared]
    public class ParameterRefactoryCodeFixProvider : CodeFixProvider
    {
        public sealed override ImmutableArray<string> FixableDiagnosticIds =>
            ImmutableArray.Create(DiagnosticId.ParameterRefactory.ToDiagnosticId());

        public sealed override FixAllProvider GetFixAllProvider() => WellKnownFixAllProviders.BatchFixer;

        public sealed override Task RegisterCodeFixesAsync(CodeFixContext context)
        {
            var diagnostic = context.Diagnostics.First();
            context.RegisterCodeFix(CodeAction.Create("Change to new Class", c => NewClassAsync(context.Document, diagnostic, c), nameof(ParameterRefactoryCodeFixProvider)), diagnostic);
            return Task.FromResult(0);
        }

        private async static Task<Document> NewClassAsync(Document document, Diagnostic diagnostic, CancellationToken cancellationToken)
        {
            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
            var diagnosticSpan = diagnostic.Location.SourceSpan;
            var oldClass = root.FindToken(diagnosticSpan.Start).Parent.AncestorsAndSelf().OfType<ClassDeclarationSyntax>().First();
            var oldNamespace = root.FindToken(diagnosticSpan.Start).Parent.AncestorsAndSelf().OfType<NamespaceDeclarationSyntax>().FirstOrDefault();
            var oldMethod = root.FindToken(diagnosticSpan.Start).Parent.AncestorsAndSelf().OfType<MethodDeclarationSyntax>().First();
            SyntaxNode newRootParameter = null;
            if (oldNamespace == null)
            {
                var newCompilation = AddParameterClassToCompilationUnitAndUpdateClassToUseNamespace((CompilationUnitSyntax)oldClass.Parent, oldClass, oldMethod);
                newRootParameter = root.ReplaceNode(oldClass.Parent, newCompilation);
                return document.WithSyntaxRoot(newRootParameter);
            }
            var newNamespace = AddParameterClassToNamespaceAndUpdateClassToUseNamespace(oldNamespace, oldClass, oldMethod);
            newRootParameter = root.ReplaceNode(oldNamespace, newNamespace).WithAdditionalAnnotations(Formatter.Annotation);
            return document.WithSyntaxRoot(newRootParameter);
        }

        private static List<PropertyDeclarationSyntax> CreateProperties(MethodDeclarationSyntax method)
        {
            var newGetSyntax = SyntaxFactory.AccessorDeclaration(SyntaxKind.GetAccessorDeclaration)
                .WithSemicolonToken(SyntaxFactory.Token(SyntaxKind.SemicolonToken));
            var newSetSyntax = SyntaxFactory.AccessorDeclaration(SyntaxKind.SetAccessorDeclaration)
                .WithSemicolonToken(SyntaxFactory.Token(SyntaxKind.SemicolonToken));
            var acessorSyntax = SyntaxFactory.AccessorList(
                SyntaxFactory.Token(SyntaxKind.OpenBraceToken),
                SyntaxFactory.List(new[] { newGetSyntax, newSetSyntax }),
                SyntaxFactory.Token(SyntaxKind.CloseBraceToken));
            var properties = new List<PropertyDeclarationSyntax>();
            foreach (ParameterSyntax param in method.ParameterList.Parameters)
            {
                var property = SyntaxFactory.PropertyDeclaration(param.Type, FirstLetteToUpper(param.Identifier.Text))
                    .WithModifiers(SyntaxFactory.TokenList(new[] { SyntaxFactory.Token(SyntaxKind.PublicKeyword) }))
                    .WithAccessorList(acessorSyntax);
                properties.Add(property);
            }
            return properties;
        }

        private static ClassDeclarationSyntax CreateParameterClass(string newNameClass, MethodDeclarationSyntax oldMethod)
        {
            var properties = CreateProperties(oldMethod);
            return SyntaxFactory.ClassDeclaration(newNameClass)
                .WithMembers(SyntaxFactory.List<MemberDeclarationSyntax>(properties))
                .WithModifiers(SyntaxFactory.TokenList(SyntaxFactory.Token(SyntaxKind.PublicKeyword)));
        }

        private static NamespaceDeclarationSyntax AddParameterClassToNamespaceAndUpdateClassToUseNamespace(NamespaceDeclarationSyntax oldNamespace, ClassDeclarationSyntax oldClass, MethodDeclarationSyntax oldMethod)
        {
            var className = $"NewClass{oldMethod.Identifier.Text}";
            var newParameterClass = CreateParameterClass(className, oldMethod);
            var newNamespace = oldNamespace.ReplaceNode(oldClass, UpdateClassToUseNewParameterClass(className, oldClass, oldMethod))
                .AddMembers(newParameterClass);
            return newNamespace;
        }

        private static ClassDeclarationSyntax UpdateClassToUseNewParameterClass(string className, ClassDeclarationSyntax classOld, MethodDeclarationSyntax methodOld)
        {
            var newParameter = SyntaxFactory.Parameter(SyntaxFactory.Identifier(FirstLetteToLower(className))).WithType(SyntaxFactory.ParseTypeName(className));
            var parameters = SyntaxFactory.ParameterList(SyntaxFactory.SeparatedList<ParameterSyntax>().Add(newParameter));
            var newMethod = methodOld.WithParameterList(parameters);
            var newClass = classOld.ReplaceNode(methodOld, newMethod);
            return newClass;
        }

        private static CompilationUnitSyntax AddParameterClassToCompilationUnitAndUpdateClassToUseNamespace(CompilationUnitSyntax oldCompilation, ClassDeclarationSyntax oldClass, MethodDeclarationSyntax oldMethod)
        {
            var className = $"NewClass{oldMethod.Identifier.Text}";
            var newParameterClass = CreateParameterClass(className, oldMethod);
            var newNamespace = oldCompilation.ReplaceNode(oldClass, UpdateClassToUseNewParameterClass(className, oldClass, oldMethod))
                .AddMembers(newParameterClass);
            return newNamespace;
        }

        private static string FirstLetteToUpper(string text) =>
            string.Concat(text.Replace(text[0].ToString(), text[0].ToString().ToUpper()));

        private static string FirstLetteToLower(string text) =>
            string.Concat(text.Replace(text[0].ToString(), text[0].ToString().ToLower()));
    }
}

File: src/VisualBasic/CodeCracker/Refactoring/ParameterRefactoryCodeFixProvider.vb
Imports System.Collections.Immutable
Imports System.Threading
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.CodeActions
Imports Microsoft.CodeAnalysis.CodeFixes
Imports Microsoft.CodeAnalysis.Formatting
Imports Microsoft.CodeAnalysis.VisualBasic
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax

Namespace Refactoring
    <ExportCodeFixProvider(LanguageNames.VisualBasic, Name:=NameOf(ParameterRefactoryCodeFixProvider)), Composition.Shared>
    Public Class ParameterRefactoryCodeFixProvider
        Inherits CodeFixProvider

        Public Overrides Function RegisterCodeFixesAsync(context As CodeFixContext) As Task
            Dim diagnostic = context.Diagnostics.First()
            context.RegisterCodeFix(CodeAction.Create("Change to new Class", Function(c) NewClassAsync(context.Document, diagnostic, c), NameOf(ParameterRefactoryCodeFixProvider)), diagnostic)
            Return Task.FromResult(0)
        End Function

        Public Overrides NotOverridable ReadOnly Property FixableDiagnosticIds As ImmutableArray(Of String) = ImmutableArray.Create(DiagnosticId.ParameterRefactory.ToDiagnosticId())

        Public Overrides Function GetFixAllProvider() As FixAllProvider
            Return WellKnownFixAllProviders.BatchFixer
        End Function

        Private Async Function NewClassAsync(document As Document, diagnostic As Diagnostic, cancellationToken As CancellationToken) As Task(Of Document)
            Dim root = Await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(False)
            Dim diagnosticSpan = diagnostic.Location.SourceSpan
            Dim declarationClass = root.FindToken(diagnosticSpan.Start).Parent.FirstAncestorOfType(Of ClassBlockSyntax)
            Dim declarationNamespace = root.FindToken(diagnosticSpan.Start).Parent.FirstAncestorOfType(Of NamespaceBlockSyntax)
            Dim declarationMethod = root.FindToken(diagnosticSpan.Start).Parent.FirstAncestorOfType(Of MethodBlockSyntax)

            Dim newRootParameter As SyntaxNode
            If declarationNamespace Is Nothing Then
                Dim newCompilation = NewCompilationFactory(DirectCast(declarationClass.Parent, CompilationUnitSyntax), declarationClass, declarationMethod)
                newRootParameter = root.ReplaceNode(declarationClass.Parent, newCompilation)
                Return document.WithSyntaxRoot(newRootParameter)
            End If
            Dim newNamespace = NewNamespaceFactory(declarationNamespace, declarationClass, declarationMethod)
            newRootParameter = root.ReplaceNode(declarationNamespace, newNamespace)
            Return document.WithSyntaxRoot(newRootParameter)
        End Function

        Private Shared Function NewPropertyClassFactory(methodOld As MethodBlockSyntax) As List(Of PropertyStatementSyntax)
            Dim properties = New List(Of PropertyStatementSyntax)
            For Each param In methodOld.SubOrFunctionStatement.ParameterList.Parameters
                Dim newProperty = SyntaxFactory.PropertyStatement(FirstLetterToUpper(param.Identifier.GetText().ToString())).
                    WithModifiers(SyntaxFactory.TokenList(SyntaxFactory.Token(SyntaxKind.PublicKeyword))).
                    WithAsClause(param.AsClause).
                    WithAdditionalAnnotations(Formatter.Annotation)

                properties.Add(newProperty)
            Next
            Return properties
        End Function

        Private Shared Function NewClassParameterFactory(newClassName As String, properties As List(Of PropertyStatementSyntax)) As ClassBlockSyntax
            Dim declaration = SyntaxFactory.ClassStatement(newClassName).
                WithModifiers(SyntaxFactory.TokenList(SyntaxFactory.Token(SyntaxKind.PublicKeyword)))

            Return SyntaxFactory.ClassBlock(declaration).
                WithMembers(SyntaxFactory.List(Of StatementSyntax)(properties)).
                WithTrailingTrivia(SyntaxFactory.CarriageReturnLineFeed).
                WithAdditionalAnnotations(Formatter.Annotation)
        End Function

        Private Shared Function NewNamespaceFactory(oldNamespace As NamespaceBlockSyntax, oldClass As ClassBlockSyntax, oldMethod As MethodBlockSyntax) As NamespaceBlockSyntax
            Dim newNamespace = oldNamespace
            Dim className = "NewClass" & oldMethod.SubOrFunctionStatement.Identifier.Text
            Dim memberNamespaceOld = oldNamespace.Members.FirstOrDefault(Function(member) member.Equals(oldClass))
            newNamespace = oldNamespace.ReplaceNode(memberNamespaceOld, NewClassFactory(className, oldClass, oldMethod))
            Dim newParameterClass = NewClassParameterFactory(className, NewPropertyClassFactory(oldMethod))
            newNamespace = newNamespace.
                WithMembers(newNamespace.Members.Add(newParameterClass)).
                WithAdditionalAnnotations(Formatter.Annotation)
            Return newNamespace
        End Function

        Private Shared Function NewClassFactory(className As String, oldClass As ClassBlockSyntax, oldMethod As MethodBlockSyntax) As ClassBlockSyntax
            Dim newParameter = SyntaxFactory.Parameter(identifier:=SyntaxFactory.ModifiedIdentifier(FirstLetterToLower(className))).
                WithAsClause(SyntaxFactory.SimpleAsClause(SyntaxFactory.IdentifierName(className))).
                NormalizeWhitespace(" ")

            Dim parameters = SyntaxFactory.ParameterList(SyntaxFactory.SeparatedList(Of ParameterSyntax).Add(newParameter)).
                WithAdditionalAnnotations(Formatter.Annotation)

            Dim methodStatement = If(oldMethod.Kind = SyntaxKind.SubBlock,
                SyntaxFactory.SubStatement(oldMethod.SubOrFunctionStatement.Identifier.Text).
                    WithModifiers(oldMethod.SubOrFunctionStatement.Modifiers).
                    WithParameterList(parameters).
                    WithTrailingTrivia(SyntaxFactory.CarriageReturnLineFeed).
                    WithAdditionalAnnotations(Formatter.Annotation),
                SyntaxFactory.FunctionStatement(oldMethod.SubOrFunctionStatement.Identifier.Text).
                    WithModifiers(oldMethod.SubOrFunctionStatement.Modifiers).
                    WithParameterList(parameters).
                    WithAsClause(oldMethod.SubOrFunctionStatement.AsClause).
                    WithTrailingTrivia(SyntaxFactory.CarriageReturnLineFeed).
                    WithAdditionalAnnotations(Formatter.Annotation))

            Dim newMethod = SyntaxFactory.MethodBlock(oldMethod.Kind, methodStatement, oldMethod.Statements, oldMethod.EndSubOrFunctionStatement).
                WithAdditionalAnnotations(Formatter.Annotation)
            Dim newClass = oldClass.ReplaceNode(oldMethod, newMethod)
            Return newClass
        End Function

        Private Shared Function FirstLetterToUpper(input As String) As String
            Return String.Concat(input.Replace(input(0).ToString(), input(0).ToString().ToUpper()))
        End Function

        Private Shared Function FirstLetterToLower(input As String) As String
            Return String.Concat(input.Replace(input(0).ToString(), input(0).ToString().ToLower()))
        End Function

        Private Shared Function NewCompilationFactory(oldCompilation As CompilationUnitSyntax, oldClass As ClassBlockSyntax, oldMethod As MethodBlockSyntax) As CompilationUnitSyntax
            Dim newNamespace = oldCompilation
            Dim className = "NewClass" & oldMethod.SubOrFunctionStatement.Identifier.Text
            Dim oldMemberNamespace = oldCompilation.Members.FirstOrDefault(Function(member) member.Equals(oldClass))
            newNamespace = oldCompilation.ReplaceNode(oldMemberNamespace, NewClassFactory(className, oldClass, oldMethod))
            Dim newParameterClass = NewClassParameterFactory(className, NewPropertyClassFactory(oldMethod))

            Return newNamespace.WithMembers(newNamespace.Members.Add(newParameterClass)).
                WithAdditionalAnnotations(Formatter.Annotation)
        End Function
    End Class
End Namespace